File: dbt_platform/engine/validators/lineage.py
# -*- coding: utf-8 -*-
"""
通用血缘计算引擎

输入: 模型文件目录
输出: DAG 节点-边 JSON
"""
import re
import json
from pathlib import Path
from typing import Optional


def compute_lineage(models_dir: str, macros_dir: Optional[str] = None) -> dict:
    """扫描模型目录，返回 DAG 结构。

    Returns:
        {
            "nodes": [{"id": "n0", "name": "dwd_order_vbrk", "layer": "DWD", "domain": "order"}],
            "edges": [{"from": "n0", "to": "n1", "type": "ref"}],
            "layer_counts": {"ODS": 74, "DWD": 87, "DWS": 8, "DIM": 1, "ADS": 1},
        }
    """
    nodes = []
    edges = []
    node_map = {}
    node_id = 0

    def add_node(name, layer, domain, path, extra=None):
        nonlocal node_id
        if name in node_map:
            return node_map[name]
        nid = f"n{node_id}"
        node_id += 1
        node = {"id": nid, "name": name, "layer": layer, "domain": domain, "path": path}
        if extra:
            node.update(extra)
        nodes.append(node)
        node_map[name] = nid
        return nid

    def add_edge(from_name, to_name, edge_type):
        fid = node_map.get(from_name)
        tid = node_map.get(to_name)
        if fid and tid and fid != tid:
            edges.append({"from": fid, "to": tid, "type": edge_type})

    models_path = Path(models_dir)
    if not models_path.exists():
        return {"nodes": [], "edges": [], "layer_counts": {}}

    for sql_file in models_path.rglob("*.sql"):
        model_name = sql_file.stem
        content = sql_file.read_text(encoding="utf-8")

        # 推断层级和域
        parts = sql_file.parts
        layer = "UNKNOWN"
        domain = "common"
        for p in parts:
            if p in ("dwd", "dws", "dim", "ads"):
                layer = p.upper()
            elif p not in ("models", "intermediate", "finance", "editorial"):
                if layer != "UNKNOWN":
                    domain = p

        # 提取 source 引用
        sources = re.findall(r"source\('ods',\s*'([^']+)'\)", content)
        for src in sources:
            add_node(src, "ODS", domain, "—", {"is_source": True})
            add_edge(model_name, src, "source")

        add_node(model_name, layer, domain,
                 str(sql_file.relative_to(models_path.parent)))

        # 提取 ref 引用
        refs = re.findall(r"ref\('([^']+)'\)", content)
        for ref in refs:
            add_edge(model_name, ref, "ref")

    layer_counts = {}
    for n in nodes:
        layer_counts[n["layer"]] = layer_counts.get(n["layer"], 0) + 1

    return {"nodes": nodes, "edges": edges, "layer_counts": layer_counts}


if __name__ == "__main__":
    import sys
    result = compute_lineage(sys.argv[1] if len(sys.argv) > 1 else "models")
    print(json.dumps(result, ensure_ascii=False, indent=2))

File: platform/electron/modules/lineage.js
/**
 * 模块2: DAG 血缘计算
 *
 * 解析所有模型的 ref() 和 source() 引用，构建完整的 DAG 节点-边图。
 * 纯函数，与 engine/validators/lineage.py 对称。
 */

const path = require('path');
const fs = require('fs');

/**
 * 计算完整血缘 DAG
 * @param {string} projectRoot — 项目根目录
 * @returns {{ nodes: object[], edges: object[], nodeMap: object, layerCounts: object }}
 */
function computeLineage(projectRoot) {
  const nodes = [];
  const edges = [];
  const nodeMap = {};
  let nodeId = 0;

  const addNode = (name, layer, domain, filePath, extra = {}) => {
    if (nodeMap[name]) return nodeMap[name];
    const id = `n${nodeId++}`;
    const node = { id, name, layer, domain, path: filePath, ...extra };
    nodes.push(node);
    nodeMap[name] = id;
    return id;
  };

  const addEdge = (fromName, toName, type) => {
    const fromId = nodeMap[fromName];
    const toId = nodeMap[toName];
    if (fromId && toId && fromId !== toId) {
      const key = `${fromId}->${toId}`;
      if (!edges.some(e => e.key === key)) {
        edges.push({ key, from: fromId, to: toId, type });
      }
    }
  };

  // Scan DWD models
  const dwdPath = path.join(projectRoot, 'models', 'dwd');
  if (fs.existsSync(dwdPath)) {
    for (const domain of fs.readdirSync(dwdPath)) {
      const dp = path.join(dwdPath, domain);
      if (!fs.statSync(dp).isDirectory()) continue;
      for (const file of fs.readdirSync(dp)) {
        if (!file.endsWith('.sql')) continue;
        const modelName = file.replace('.sql', '');
        const content = fs.readFileSync(path.join(dp, file), 'utf-8');
        const sourceMatch = content.match(/\{\{\s*source\('ods',\s*'([^']+)'\)\s*\}\}/);
        const sourceName = sourceMatch ? sourceMatch[1] : null;
        addNode(modelName, 'DWD', domain, `models/dwd/${domain}/${file}`, {
          sourceTable: sourceName,
          isIncremental: content.includes('is_incremental'),
          hasDedup: content.includes('ROW_NUMBER'),
        });
        const refs = content.match(/\{\{\s*ref\('([^']+)'\)\s*\}\}/g);
        if (refs) {
          for (const r of refs) {
            const refName = r.match(/ref\('([^']+)'\)/)[1];
            addEdge(modelName, refName, 'ref');
          }
        }
      }
    }
  }

  // Scan DWS / DIM / ADS (recursive)
  const scanModels = (dirPath, layer, domain = null) => {
    if (!fs.existsSync(dirPath)) return;
    for (const entry of fs.readdirSync(dirPath, { withFileTypes: true })) {
      if (entry.isDirectory()) {
        scanModels(path.join(dirPath, entry.name), layer, domain || entry.name);
      } else if (entry.name.endsWith('.sql')) {
        const modelName = entry.name.replace('.sql', '');
        const content = fs.readFileSync(path.join(dirPath, entry.name), 'utf-8');
        addNode(modelName, layer, domain || 'common', path.relative(projectRoot, path.join(dirPath, entry.name)));
        const refs = content.match(/\{\{\s*ref\('([^']+)'\)\s*\}\}/g);
        if (refs) {
          for (const r of refs) {
            const refName = r.match(/ref\('([^']+)'\)/)[1];
            addEdge(modelName, refName, 'ref');
          }
        }
      }
    }
  };

  scanModels(path.join(projectRoot, 'models', 'dws'), 'DWS');
  scanModels(path.join(projectRoot, 'models', 'dim'), 'DIM');
  scanModels(path.join(projectRoot, 'models', 'ads'), 'ADS');

  // Add ODS source nodes
  for (const node of nodes) {
    if (node.sourceTable) {
      const odsName = `ods_sap_${node.sourceTable}`;
      addNode(odsName, 'ODS', node.domain || 'unknown', '—', { isSource: true });
      addEdge(node.name, odsName, 'source');
    }
  }

  // Layer counts
  const layerCounts = {};
  for (const n of nodes) {
    layerCounts[n.layer] = (layerCounts[n.layer] || 0) + 1;
  }

  return { nodes, edges, nodeMap, layerCounts };
}

module.exports = { computeLineage };
